tests/fixture/bootstrapmonitor/common.go

```diff
@@ -0,0 +1,294 @@
+	}
+	log.Info("Updated statefulset to target new image",
+		zap.String("namespace", namespace),
+		zap.String("statefulSetName", statefulSetName),
+		zap.String("image", imageDetails.Image),
+		zap.Reflect("versions", imageDetails.Versions),
+	)
+	return nil
+}
+
+// getBaseImageName removes the tag from the image name
+func getBaseImageName(log logging.Logger, imageName string) (string, error) {
+	if strings.Contains(imageName, "@") {
+		// Image name contains a digest, remove it
+		return strings.Split(imageName, "@")[0], nil
+	}
+
+	imageNameParts := strings.Split(imageName, ":")
+	switch len(imageNameParts) {
+	case 1:
+		// No tag or registry
+		return imageName, nil
+	case 2:
+		// Ambiguous image name - could contain a tag or a registry
+		log.Info("Derived tag-less image name from string",
+			zap.String("tagLessImageName", imageNameParts[0]),
+			zap.String("imageName", imageName),
+		)
+		return imageNameParts[0], nil
+	case 3:
+		// Image name contains a registry and a tag - remove the tag
+		return strings.Join(imageNameParts[0:2], ":"), nil
+	default:
+		return "", fmt.Errorf("unexpected image name format: %q", imageName)
+	}
+}
+
+type ImageDetails struct {
+	Image    string
+	Versions *version.Versions
+}
+
+// GetLatestImageDetails retrieves the image details for the avalanchego image with tag `latest`.
+func getLatestImageDetails(
+	ctx context.Context,
+	log logging.Logger,
+	clientset *kubernetes.Clientset,
+	namespace string,
+	imageName string,
+	containerName string,
+) (*ImageDetails, error) {
+	baseImageName, err := getBaseImageName(log, imageName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start a new pod with the `latest`-tagged avalanchego image to discover its image ID
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "avalanchego-version-check-",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:    containerName,
+					Command: []string{"./avalanchego"},
+					Args:    []string{"--version-json"},
+					Image:   baseImageName + ":latest",
+				},
+			},
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+	createdPod, err := clientset.CoreV1().Pods(namespace).Create(ctx, pod, metav1.CreateOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to start pod %w", err)
+	}
+	qualifiedPodName := fmt.Sprintf("%s.%s", namespace, createdPod.Name)
+
+	err = waitForPodStatus(ctx, clientset, namespace, createdPod.Name, func(status *corev1.PodStatus) bool {
+		return status.Phase == corev1.PodSucceeded || status.Phase == corev1.PodFailed
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for pod %s to terminate: %w", qualifiedPodName, err)
+	}
+
+	terminatedPod, err := clientset.CoreV1().Pods(namespace).Get(ctx, createdPod.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve terminated pod %s: %w", qualifiedPodName, err)
+	}
+
+	// Get the image id for the avalanchego image
+	imageID := ""
```

Comment:
I've seen that you have been using this pattern quite frequently within this PR. maybe consider to refactor in some way ? ( future PR )

Reply:
sure. I'm referring to the pattern you've used here:

Reply:
I've experimented with several permutation for the above; I didn't really like the Index here, since it's not the index that is being looked for.
```golang
        func SelectSliceElement[K any, V any](slc []K, fnc func(e *K) (V, bool), miss func() error) (V, error) {
	        for _, e := range slc {
		        if r, found := fnc(&e); found {
			        return r, nil
		        }
	        }
	        var v V
	        return v, miss()
        }

...
	// Get the image id for the avalanchego image
	imageID, err := SelectSliceElement[corev1.ContainerStatus, string](terminatedPod.Status.ContainerStatuses,
		func(e *corev1.ContainerStatus) (string, bool) {
			return e.ImageID, (e.Name == containerName) && len(containerName) > 0
		},
		func() error {
			return fmt.Errorf("failed to get image id for pod %s", qualifiedPodName)
		})
	if err != nil {
		return nil, err
	}
```

but that doesn't seem to be right. It's definitely more structured, but not necessarily simpler or cleaner.

What I ended up with was this:
```golang
        func FirstContainerStatusByName(slc []corev1.ContainerStatus, name string) corev1.ContainerStatus {
	        for _, status := range slc {
		        if status.Name == name {
			        return status
		        }
	        }
	        return corev1.ContainerStatus{}
        }

...

	// Get the image id for the avalanchego image
	imageID := FirstContainerStatusByName(terminatedPod.Status.ContainerStatuses, containerName).ImageID
	if len(imageID) == 0 {
		return nil, fmt.Errorf("failed to get image id for pod %s", qualifiedPodName)
	}
```

I think that it's a cleaner solution, at the expense of writing the `FirstContainerStatusByName`.
( but please don't let this block your PR )